### PR TITLE
refactor(api): single-source the typed public API overload signatures (#903)

### LINKED ISSUE #903: TIDY_UP-10: typed_public_api overload signatures need one owner
## Files
- `src/climate_indices/typed_public_api.py` — overloads and wrappers (:76-575), decorator config (:44-67), manual PCI wrapper (:330-390)
- `src/climate_indices/__init__.py:16-24` — re-exports this module as the package API
- `docs/adr/0001-dual-numpy-xarray-api.md`

## Problem
Every public function repeats its full signature up to three times: two `@overload` stubs plus the implementation (e.g. `spi` :76-159, `spei` :160-255), and the implementation only filters kwargs and delegates (`return _wrapped_*`). Signatures can drift from `indices.py` with nothing catching it (back-compat tests freeze only a few via `inspect` — `tests/test_backward_compat.py:430-480`). PCI needs a manual wrapper because its output shape differs, adding a shape no other index uses.

## Solution
Pick one owner:
- generate the overloads from the `indices` signatures (codegen or a `ParamSpec`-based helper), or
- make `typed_public_api` the single owner of the decorator config and derive signatures from one source.

Keep ADR-0001's constraint: dual API stays, one computation source of truth, xarray wraps numpy.

## Benefits
Locality: one signature per public function, checked by mypy against `indices`. Leverage: adding a parameter stops being a three-place edit.

## Recommendation strength
**Worth exploring**

## Related
#847, #856 (closed), ADR-0001, #885

## Acceptance
- Public signatures and outputs unchanged; mypy still narrows ndarray vs DataArray overloads.
- `uv run ruff check src/ tests/`, `uv run ruff format --check src/ tests/`, `uv run mypy src/`, `uv run pytest` pass.

### PR DESCRIPTION
**Pull request by monocongo**
## Summary

The pass-through wrappers in `typed_public_api` each declared their full signature three times: two `@overload` stubs plus the implementation. The implementation copy could drift from the stubs and from the `indices.py` / `xarray_adapter` function it delegates to, with nothing catching it.

This PR removes the third copy while keeping the wrappers' observable behavior:

- `_delegate(func, data, *args, **kwargs)` binds the call against the delegate's signature, forwards the data positionally (the adapter requires `args[0]`) and the rest as keywords, rejects unknown keyword arguments with `TypeError`, and drops explicit `None` only for parameters without a default — binding `None` would suppress the adapter's inference of that parameter.
- Each implementation is now e.g. `def spi(values, *args, **kwargs): return _delegate(_wrapped_spi, values, *args, **kwargs)`.
- `_restore_runtime_signature` pins each public function's `inspect`/Sphinx signature to its NumPy `@overload` stub (needs `typing.get_overloads`, so Python 3.11+; the docs gate runs 3.12), keeping the API reference's parameter list.
- Tests freeze both overload signatures, pin their parameter names to the wrapped function, and cover positional + Dask-positional forwarding, unknown-keyword rejection, explicit-`None` equivalence, and the overload table's coverage of the public API.

## Scope decisions

- **No codegen.** The issue offered generating the overloads from the `indices` signatures. Rejected: a generator plus a CI drift gate is more machinery than the ~100 lines of overloading it would remove, for seven functions.
- **Overload stubs stay hand-written.** Python cannot derive a narrowed `@overload` pair from a runtime signature, and the module's contract — strict mypy narrowing plus IDE autocomplete for the two dialects — depends on those stubs. A `ParamSpec`-typed `xarray_adapter` would derive the NumPy half but cannot express the xarray half's optional parameters, so it does not remove the stubs either.
- **Displayed signature form.** `inspect`/Sphinx now show the NumPy overload, so the temporal parameters render as required (the NumPy contract) rather than the old union's `= None`; every docstring already states that they are optional and inferred for xarray input.
- **`pci` keeps its explicit implementation signature**: real body logic (scalar output shape, CF metadata), not a pass-through.
- **`spatial_time_major`** stays out of the public overloads (it is the adapter's block-protocol keyword), but generic forwarding now lets it reach the NumPy path where the explicit signatures raised `TypeError`; this is a deliberate widening of a valid `indices.*` parameter, not advertised by the typed API.

## Verification

On the merged HEAD (includes main through #1022):

- `uv run ruff check src/ tests/` — pass
- `uv run ruff format --check src/ tests/` — pass
- `uv run mypy src/ tests/test_type_checking.py` — pass
- `uv run mypy --strict src/climate_indices/typed_public_api.py` — pass
- `uv run pytest` — 2151 passed, 423 deselected
- `uv run sphinx-build -E -b html -W --keep-going docs docs/_build/html` — pass
- `uv run sphinx-build -E -b doctest docs docs/_build/doctest` — pass

Closes #903

## Summary by Sourcery

Refactor typed public API wrappers to eliminate duplicate implementation signatures while preserving their typing contract, runtime behavior, and documented interfaces.

Bug Fixes:
- Preserve pass-through API behavior for positional arguments, Dask execution, unknown-keyword rejection, and explicit `None` inference handling.

Enhancements:
- Single-source typed public API wrappers by retaining hand-written NumPy and xarray overloads while using generic implementations and shared delegation.
- Restore runtime and documentation signatures from the NumPy overloads and add safeguards against overload drift and incomplete API coverage.

Documentation:
- Keep the generated API reference aligned with the published NumPy overload signatures.

Tests:
- Add coverage for overload contracts, delegate forwarding behavior, invalid keywords, and `None` handling.

## Summary by CodeRabbit

- **Improvements**
  - Updated the public climate index API for SPI, SPEI, Percentage of Normal, Thornthwaite PET, Hargreaves PET, and EDDI.
  - Preserved function signatures for static type checking and runtime inspection.
  - Standardized positional and keyword argument handling across NumPy, xarray, Dask, and explicit `None` inputs.
  - Invalid or misspelled keyword arguments now raise clear `TypeError` exceptions.

- **Tests**
  - Added coverage for public overload signatures and argument-forwarding behavior.

### CODE REVIEW

File: src/climate_indices/typed_public_api.py [hunk 1]
```diff
@@ -69,9 +74,17 @@
...
+    func: Callable[..., npt.NDArray[np.float64] | xr.DataArray],
+    *args: Any,
+    **kwargs: Any,
+) -> npt.NDArray[np.float64] | xr.DataArray:
+    """Call a pre-built API function, dropping None kwargs so absent params are inferred.
+
+    Passing an explicit None would bind the parameter and suppress the xarray
+    adapter's inference, so the public wrappers drop them here.
+    """
+    return func(*args, **{key: value for key, value in kwargs.items() if value is not None})
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/typed_public_api.py]: Addressed in b2fcc3c6 (PR head now 16371c87).

`_delegate` no longer forwards raw `*args`: it binds the call against the wrapped function's signature, re-forwards the data as the first positional argument and every other parameter as a keyword, and drops explicit `None` only for parameters without a default (the inferred temporals). Positional `scale`/`distribution` therefore reach the adapter on the multi-dimensional/Dask `apply_ufunc` paths with their calculation metadata intact, and positional `None` temporal placeholders still fall through to inference.

Regression tests: `test_spi_xarray_positional_arguments_match_keywords`, `test_spi_dask_positional_arguments_match_keywords`, and `test_explicit_none_matches_omitted` in `tests/test_typed_public_api.py` (values and metadata compared against the keyword calls).

File: src/climate_indices/typed_public_api.py [hunk 2]
```diff
@@ -101,16 +114,7 @@ def spi(
...
-    values: npt.NDArray[np.float64] | xr.DataArray,
-    scale: int,
-    distribution: Distribution,
-    data_start_year: int | None = None,
-    calibration_year_initial: int | None = None,
-    calibration_year_final: int | None = None,
-    periodicity: Periodicity | None = None,
-    fitting_params: dict[str, Any] | None = None,
-) -> npt.NDArray[np.float64] | xr.DataArray:
+def spi(values: Any, *args: Any, **kwargs: Any) -> npt.NDArray[np.float64] | xr.DataArray:
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/typed_public_api.py:150]: Addressed in b2fcc3c6 (PR head now 16371c87).

`_delegate` now runs `inspect.signature(func).bind_partial(data, *args, **kwargs)` before the call, so a keyword outside the wrapped signature raises `TypeError: got an unexpected keyword argument ...` instead of being silently filtered away by the adapter and replaced with an inferred value. Omitted parameters still infer, so the omission-driven contract is unchanged.

Regression test: `test_unknown_keyword_argument_is_rejected` (misspelled `calibraton_year_initial`) in `tests/test_typed_public_api.py`.

File: src/climate_indices/typed_public_api.py [hunk 3]
```diff
@@ -24,7 +30,10 @@
 from __future__ import annotations
 
 import datetime
-from typing import Any, overload
+import inspect
+import typing
+from collections.abc import Callable
+from typing import Any, cast, overload
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/typed_public_api.py:36]: Addressed in 2279672a (PR head now 6109cad8).

Python 3.10 has no `typing.get_overloads`, and stubs decorated with `typing.overload` are not registered in the runtime overload registry before 3.11. `typing_extensions.get_overloads` only tracks overloads created with `typing_extensions.overload`, so it returned an empty list on 3.10 and the three introspection tests failed there.

The resolution keeps the module on the stdlib `typing` imports and skips the three registry-dependent tests below 3.11 instead of adding a conditional `typing-extensions` runtime dependency; they run unchanged on 3.11+. `_restore_runtime_signature` keeps its documented pre-3.11 early return, and mypy still checks the overload stubs statically on 3.10.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/1025
- **Lanes run:** 5 — correctness/edge cases, public API contracts, tests, maintainability, documentation. The two Codex-model lanes hit a provider usage limit and were rerun on the alternate provider (spec/contracts and docs) before synthesizing.
- **Validation:** `uv run ruff check src/ tests/` pass; `uv run ruff format --check src/ tests/` pass; `uv run mypy src/ tests/test_type_checking.py` pass; `uv run mypy --strict src/climate_indices/typed_public_api.py` pass; `uv run pytest` 2151 passed, 423 deselected; `uv run sphinx-build -E -b html -W --keep-going docs docs/_build/html` pass; `uv run sphinx-build -E -b doctest docs docs/_build/doctest` pass.
- **Fix commits:** [b2fcc3c6 — fix(api): keep the explicit wrappers' call semantics in the generic delegate (#903)](https://github.com/monocongo/climate_indices/commit/b2fcc3c6)

### Findings and fix status

#### [P1] Positional non-data arguments dropped on the Dask path — Fixed
- **Location:** `src/climate_indices/typed_public_api.py` (was `_delegate`, now lines 81–103); `src/climate_indices/xarray_adapter.py:1500-1513`
- **Origin:** Correctness/edge cases; contracts
- **Problem:** `spi(chunked_da, 6, Distribution.gamma)` forwarded `scale`/`distribution` as positionals, but the Dask branch builds `valid_kwargs` from keywords only, so `.compute()` raised `TypeError: spi() missing 2 required positional arguments`. Even on the in-memory path the same call lost `scale`, `distribution`, and calibration-year output metadata, and skipped the calibration NaN-sample validation. The explicit signatures on `main` always forwarded those parameters as keywords.
- **Applied fix:** `_delegate` now binds the call against the delegate's signature and re-forwards everything except the data as keywords. Regression tests cover positional and Dask-positional calls, including the calculation metadata. Commit `b2fcc3c6`.

#### [P2] Misspelled/unknown keyword arguments silently ignored on the xarray path — Fixed
- **Location:** `src/climate_indices/typed_public_api.py:81-103`; `src/climate_indices/xarray_adapter.py:1574-1581`
- **Origin:** Correctness/edge cases; tests; documentation; contracts
- **Problem:** The generic `**kwargs` forwarding let `spi(da, scale=6, distribution=..., calibraton_year_initial=1981)` reach the adapter, whose DataArray path filters unknown keys. It computed over the inferred 1980–2019 calibration instead of raising, while the NumPy path still raised. The explicit signatures raised `TypeError` for both.
- **Applied fix:** `signature.bind_partial(...)` now rejects unknown keywords with `TypeError`; a regression test asserts the failure. Commit `b2fcc3c6`.

#### [P2] Explicit `None` semantics changed for positional arguments and `time_dim` — Fixed
- **Location:** `src/climate_indices/typed_public_api.py:81-103`, `:150`, `:216`, `:285`
- **Origin:** Correctness/edge cases; maintainability; contracts
- **Problem:** Dropping every `None` kwargs-only broke `spi(da, 6, dist, None)` (bound `None` suppressed inference and failed later), and also dropped `time_dim=None` for the PET wrappers, silently replacing the previous `CoordinateValidationError` with the default `"time"`. Defaulted parameters such as `fitting_params` also changed from passed-`None` to omitted.
- **Applied fix:** `None` is dropped only for parameters without a default (the inferred temporals); positional `None` is bound before the check, so `spi(da, 6, dist, None)` infers as before. Regression test covers keyword-`None` and positional-`None` equivalence with the omitted call. Commit `b2fcc3c6`.

#### [P2] `inspect`/Sphinx signature lost its parameter list — Fixed
- **Location:** `src/climate_indices/typed_public_api.py:108-123`, `:551-552`
- **Origin:** Documentation; contracts
- **Problem:** With generic implementations, the API reference rendered `spi(values, *args, **kwargs)` and `help()`/`inspect.signature()` showed the generic form, regressing against the issue's "public signatures unchanged" criterion.
- **Applied fix:** `_restore_runtime_signature` pins each public function's `__signature__` to its NumPy overload stub (needs `typing.get_overloads`, i.e. Python 3.11+; the docs gate runs 3.12). The reference again lists every parameter and the docs build is warning-free. Commit `b2fcc3c6`.

#### [P2] Overload drift test pinned parameter names only, and its table was opt-in — Fixed
- **Location:** `tests/test_typed_public_api.py:100-181`
- **Origin:** Tests; maintainability; contracts
- **Problem:** Because the implementations are now `Any`-typed, mypy no longer cross-checks stub annotations; the name-only assertion let annotation, default, and parameter-kind drift through (demonstrated with mutations such as `time_dim: str = "time"` → `"days"`), and a new overloaded public function would not be covered.
- **Applied fix:** Added a frozen expectation of both rendered overload signatures per public function (`_EXPECTED_OVERLOADS`, a deliberate tripwire) and a test asserting every overloaded function exported from `climate_indices` appears in `_PUBLIC_IMPLEMENTATIONS`. Commit `b2fcc3c6`.

#### [P3] Module docstring inaccurate after the change — Fixed
- **Location:** `src/climate_indices/typed_public_api.py:10-16`
- **Origin:** Maintainability; documentation
- **Problem:** The new paragraph claimed every public function follows the generic pattern (`pci` does not), named a nonexistent `pnp` entry point, and overstated the sync test as pinning "both stubs to the wrapped function" when it pins names.
- **Applied fix:** Scoped the claim to all functions except `pci`, corrected `pnp` → `percentage_of_normal`, and described the runtime signature pin and the frozen stubs. Commit `b2fcc3c6`.

#### [P3] `spatial_time_major` is now reachable through the public wrappers — Fixed (documented)
- **Location:** `src/climate_indices/typed_public_api.py:150`, `:216`, `:285`
- **Origin:** Contracts
- **Problem:** The explicit signatures raised `TypeError` for the adapter's block-protocol keyword; generic forwarding now passes it to the NumPy path. It is a valid `indices.*` parameter, but it stays out of the public overloads, so the two surfaces disagreed.
- **Applied fix:** Kept out of the public overloads and recorded in the PR body as a deliberate widening, not an advertised parameter. An explicit rejection would hard-code the adapter-only name in the wrapper; left for a follow-up if the maintainer prefers the old error.

#### [P3] Pre-existing docs inaccuracies — Blocked (out of scope, tracked)
- **Location:** `docs/project-overview.md:86-98`, `docs/architecture.md:111`
- **Origin:** Documentation
- **Problem:** The recommended `ci.spi(...)` example passes `distribution="gamma"` (rejected) and calibration kwargs that are not `spi` parameters; `architecture.md` claims keyword-only enforcement that does not exist. Both predate this PR, and the example now fails more quietly for the misspelled kwargs.
- **Applied fix:** **Blocked by scope** (docs refresh work, not the signature refactor); tracked as #1027 and added to the Documentation Refresh board, with the `architecture.md` claim noted there.

### Remaining blockers
None. Every verified P0–P3 finding is fixed or, for the pre-existing docs inaccuracies, tracked in #1027; no failing checks remain on the final HEAD.
